File: src/lib/storage/dynamo-stack/schema/dynamo-schemas.ts
export enum DynamoTables {
  USERS = "UserTable",
  QUESTION_BANK = "QuestionBankTable",
  ASSESSMENTS = "AssessmentsTable",
  ASSESSMENT_QUESTION_LOCATOR = "AssessmentQuestionLocatorTable",
  METRICS = "MetricsTable",
}

// USER TABLE___________________________________________________________
export type UserRecord = {
  userId: string;
  firstName: string;
  lastName: string;
  email: string;
  createdAt: string;
  executionSocketUrl: string | null;
  chatSocketUrl: string | null;
  currentAssessmentTimestamp: string | null;
};

// Example
// let user: UserRecord = {}

// QUESTION BANK TABLE__________________________________________________
export enum Difficulty {
  EASY = "easy",
  MEDIUM = "medium",
  HARD = "hard",
}

export enum Topic {
  ARRAY = "Array",
  STRING = "String",
  HASH_TABLE = "Hash Table",
  DYNAMIC_PROGRAMMING = "Dynamic Programming",
  SORTING = "Sorting",
  GREEDY = "Greedy",
  DEPTH_FIRST_SEARCH = "Depth-First Search",
  BINARY_SEARCH = "Binary Search",
  MATRIX = "Matrix",
  BREADTH_FIRST_SEARCH = "Breadth-First Search",
  TREE = "Tree",
  BIT_MANIPULATION = "Bit Manipulation",
  TWO_POINTERS = "Two Pointers",
  HEAP = "Heap (Priority Queue)",
  BINARY_TREE = "Binary Tree",
  STACK = "Stack",
  GRAPH = "Graph",
  SLIDING_WINDOW = "Sliding Window",
  LINKED_LIST = "Linked List",
  ORDERED_SET = "Ordered Set",
  QUEUE = "Queue",
  RECURSION = "Recursion",
}

export enum Language {
  PYTHON = "python",
  JAVASCRIPT = "javascript",
}

export type QuestionRecord = {
  questionId: string;
  topics: Topic[]; // Array of Topic enums
  difficulty: Difficulty;
  title: string;
  description: string;
  testCases: string[];
  testAnswers: string[];
  ststarterCode: Partial<Record<Language, string>>;
  hints: string[];
};

// METRICS TABLE_________________________________________________________
export enum MetricTypes {
  ATTEMPTS = "attempts",
  EXECUTION_TIME = "execution_time",
  EXECUTION_MEMORY = "execution_memory",
  TIME_SPENT = "time_spent",
}

export type Metrics = Record<
  MetricTypes,
  {
    [scope: string]: {
      count: number;
      total: number;
    };
  }
>;

export type MetricsRecord = {
  userId: string;
  metrics: Metrics;
};

// ASSESSMENTS TABLE____________________________________________________
export enum Status {
  COMPLETE = "complete",
  ONGOING = "ongoing",
  INCOMPLETE = "incomplete",
}

export enum AttemptStatus {
  SUCCESS = "success",
  FAIL = "fail",
}

export enum QuestionStatus {
  PASS = "pass",
  INCOMPLETE = "incomplete",
  FAIL = "fail",
}

export type Attempt = {
  execTimeTaken: number; // Time taken in seconds
  execMemoryTaken: number; // Memory taken in KB
  status: AttemptStatus; // Status of the attempt (success or fail)
};

export type QuestionsDone = {
  questionId: string;
  attempts: Attempt[];
  timeStarted: string;
  timeEnded: string;
  bestExecTime: number;
  bestExecMem: number;
  testCasesPassed: number;
  status: QuestionStatus;
};

export type AssessmentRecord = {
  userId: string; // User ID as partition key
  timestamp: string; // Timestamp (as sort key)
  selectedTopics: Topic[];
  selectedDifficulty: Difficulty[];
  selectedDuration: number; // Duration in minutes
  selectedNumberOfQuestions: number;
  status: Status;
  metrics: Metrics;
  questions: QuestionsDone[];
};

// ASSESSMENT QUESTION LOCATOR TABLE____________________________________
export type LocationEntry = {
  [assessmentTimestamp: string]: string[]; // assessmentTimestamp -> array of indices
};

export type QuestionLocatorRecord = {
  userId: string; // User ID as partition key
  scope: string; // Scope as sort key
  location: LocationEntry;
};

File: infrastructure/lib/executor/result-manager-stack/lambda-code/dynamo_schemas.py
from dataclasses import dataclass
from enum import Enum
from typing import List, Dict, Optional


# TABLE NAMES___________________________________________________________
class DynamoTables(Enum):
    USERS = ("UserTable",)
    QUESTION_BANK = "QuestionBankTable"
    ASSESSMENTS = "AssessmentsTable"
    ASSESSMENT_QUESTION_LOCATOR = "AssessmentQuestionLocatorTable"
    METRICS = "MetricsTable"


# USER TABLE___________________________________________________________
@dataclass
class UserRecord:
    userId: str
    firstName: str
    lastName: str
    email: str
    createdAt: str
    executionSocketUrl: Optional[str] = None
    chatSocketUrl: Optional[str] = None
    currentAssessmentTimestamp: Optional[str] = None


# user: UserRecord = UserRecord()


# QUESTION BANK TABLE__________________________________________________
class Difficulty(Enum):
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


class Topic(Enum):
    ARRAY = "Array"
    STRING = "String"
    HASH_TABLE = "Hash Table"
    DYNAMIC_PROGRAMMING = "Dynamic Programming"
    SORTING = "Sorting"
    GREEDY = "Greedy"
    DEPTH_FIRST_SEARCH = "Depth-First Search"
    BINARY_SEARCH = "Binary Search"
    MATRIX = "Matrix"
    BREADTH_FIRST_SEARCH = "Breadth-First Search"
    TREE = "Tree"
    BIT_MANIPULATION = "Bit Manipulation"
    TWO_POINTERS = "Two Pointers"
    HEAP = "Heap (Priority Queue)"
    BINARY_TREE = "Binary Tree"
    STACK = "Stack"
    GRAPH = "Graph"
    SLIDING_WINDOW = "Sliding Window"
    LINKED_LIST = "Linked List"
    ORDERED_SET = "Ordered Set"
    QUEUE = "Queue"
    RECURSION = "Recursion"


class Language(Enum):
    PYTHON = "python"
    JAVASCRIPT = "javascript"


@dataclass
class QuestionRecord:
    questionId: str
    topics: List[Topic]  # Array of Topic enums
    difficulty: Difficulty
    title: str
    description: str
    testCases: List[str]
    testAnswers: List[str]
    starterCode: Dict[Language, str]
    hints: List[str]


# METRICS TABLE_________________________________________________________
class MetricTypes(Enum):
    ATTEMPTS = "attempts"
    EXECUTION_TIME = "execution_time"
    EXECUTION_MEMORY = "execution_memory"
    TIME_SPENT = "time_spent"


@dataclass
class MetricsValues:
    count: int
    total: int


@dataclass
class MetricScopes:
    scopes: Dict[str, MetricsValues]


@dataclass
class Metrics:
    metrics: Dict[MetricTypes, MetricScopes]


@dataclass
class MetricsRecord:
    userId: str
    metrics: Metrics


# ASSESSMENTS TABLE____________________________________________________
class Status(Enum):
    COMPLETE = "complete"
    ONGOING = "ongoing"
    INCOMPLETE = "incomplete"


class AttemptStatus(Enum):
    SUCCESS = "success"
    FAIL = "fail"


class QuestionStatus(Enum):
    PASS = "pass"
    INCOMPLETE = "incomplete"
    FAIL = "fail"


@dataclass
class Attempt:
    execTimeTaken: int  # Time taken in seconds
    execMemoryTaken: int  # Memory taken in KB
    status: AttemptStatus  # Status of the attempt (success or fail)


@dataclass
class QuestionsDone:
    questionId: str
    attempts: List[Attempt]
    timeStarted: str
    timeEnded: str
    bestExecTime: int
    bestExecMem: int
    testCasesPassed: int
    status: QuestionStatus


@dataclass
class AssessmentRecord:
    userId: str  # User ID as partition key
    timestamp: str  # Timestamp (as sort key)
    selectedTopics: List[Topic]
    selectedDifficulty: List[Difficulty]
    selectedDuration: int  # Duration in minutes
    selectedNumberOfQuestions: int
    status: Status
    metrics: Metrics
    questions: List[QuestionsDone]


# ASSESSMENT QUESTION LOCATOR TABLE____________________________________
@dataclass
class LocationEntry:
    # assessmentTimestamp -> array of indices
    assessmentTimestamp: Dict[str, List[str]]


@dataclass
class QuestionLocatorRecord:
    userId: str  # User ID as partition key
    scope: str  # Scope as sort key
    location: LocationEntry
